pytorch_lightning/core/lightning.py

```diff
@@ -1757,3 +1700,49 @@ def get_tqdm_dict(self) -> Dict[str, Union[int, str]]:
         rank_zero_warn("`get_tqdm_dict` was renamed to `get_progress_bar_dict` in v0.7.3"
                        " and this method will be removed in v1.0.0", DeprecationWarning)
         return self.get_progress_bar_dict()
+
+    def _auto_collect_arguments(self):
+        """Collect all arguments module arguments."""
+        frame = inspect.currentframe()
+
+        frame_args = _collect_init_args(frame.f_back, [])
+        child = _get_latest_child(frame)
+
+        # set module_arguments in child
+        child._module_self_arguments = frame_args[-1]
+        child._module_parents_arguments = {}
+        for args in frame_args[:-1]:
+            child._module_parents_arguments.update(args)
+
+    @property
+    def module_arguments(self) -> dict:
+        """Aggregate this module and all parents arguments."""
+        args = dict(self._module_parents_arguments)
+        args.update(self._module_self_arguments)
+        return args
+
+
+def _collect_init_args(frame, path_args: list) -> list:
+    """Recursive search for all children."""
+    if '__class__' in frame.f_locals:
```

Comment:
What does this do? 
Here is a an example we should try to handle or at least give a warning
```python
from pytorch_lightning import LightningModule
class Example(LightningModule):

    def __init__(this, arg):
        super().__init__()
        this.arg = arg

    def forward(self, x):
        pass

x = Example(1)
print(x.module_arguments)
#  {'this': Example(), 'arg': 1}
```
module_arguments contains the object itself. 
If this the PR needs to be merged asap I hope we can put it at least on the list of TODOs.

Reply:
true, it would contain the aka `self` but it will be filtered out as it is not a primitive...
so you recommend doing the primitive filtering already here?

Reply:
not sure, but we need to also inspect the constructor to get a list of accepted args and filter based on that, because of this example:
from pytorch_lightning import LightningModule

```python
class Example(LightningModule):

    def __init__(self, arg):
        my_local_var = 2
        super().__init__()

    def forward(self, x):
        pass

x = Example(1)
print(x.module_arguments)
# {'arg': 1, 'my_local_var': 2}
```
This will fail when we try to restore and pass in the local var which is not an argument in the constructor.

We should probably filter based on
`inspect.signature(Example).parameters`
and only save these locals, not the others.

Reply:
```
x = Example(1)
print(x.module_arguments)
# {'arg': 1, 'my_local_var': 2}
```
this works as in the frame it appears with arg name always...
and as it is saved you can call `x = Example(arg=1)`

Reply:
but there is an extra arg. 
`Example.load_from_checkpoint(...)`
will load module argumets 
`module_arguments = {'arg': 1, 'my_local_var': 2}`
and then call
`Example(**module_arguments) `
Which means 
my_local_var is passed in but this name is not accepted.
Thus yielding a `TypeError: __init__ got an unexpected argument "my_local_var"`
I am 99% certain.

Reply:
these are all edge cases, I just want to make everyone aware of it because I am testing this branch and trying different things. 
We can fix these in follow up PRs

Reply:
sure, good point, just to clarify that it is going to Python, not to parsed arguments
meaning you have to arr for each level `*args, **kwargs`

Reply:
well, wait if `my_local_var` is not among `__init__` arguments, it won't be present in `module_arguments` neither...

Reply:
@awaelchli mind adding an edge case to the test so we know the exact case and also we can truly test it... :]

Reply:
in follow up or here?  because william wants to merge asap he said

Reply:
@williamFalcon ^^

Reply:
let’s add it as a follow up. 

ready to merge? happy to merge now